### PR TITLE
Business logic for creating sign document

### PR DESCRIPTION
**Pull request by tba76**
## Related Issue(s)
- #211 

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

### CODE REVIEW

File: src/Storage/Controllers/SignController.cs [hunk 1]
```diff
@@ -41,8 +40,20 @@ public SignController(IInstanceService instanceService)
         [Produces("application/json")]
         public async Task<ActionResult> Sign([FromRoute] int instanceOwnerPartyId, [FromRoute] Guid instanceGuid, [FromBody] SignRequest signRequest)
         {
-            await _instanceService.CreateSignDocument(instanceOwnerPartyId, instanceGuid, signRequest);
-            return StatusCode(201, "SignDocument is created");
+            try 
+            {
+                if (string.IsNullOrEmpty(signRequest.Signee.UserId))
+                {
+                    return Problem("The 'UserId' parameter must be defined in context.", null, 400);
```
Comment (acn-sbuad) [inline src/Storage/Controllers/SignController.cs:47]:
```suggestion
                    return Problem("The 'UserId' parameter must be defined for signee.", null, 400);
```

File: src/Storage/Controllers/SignController.cs [hunk 2]
```diff
@@ -41,8 +40,20 @@ public SignController(IInstanceService instanceService)
...
+                }
+
+                await _instanceService.CreateSignDocument(instanceOwnerPartyId, instanceGuid, signRequest);
+                return StatusCode(201, "SignDocument is created");
+            } 
+            catch (Exception e)
+            {
+                return StatusCode(500, e.Message);
```
Comment (acn-sbuad) [inline src/Storage/Controllers/SignController.cs:55]: Could this also be a Problem return?

File: src/Storage/Services/ApplicationService.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ApplicationService"/> class.
+        /// </summary>
+        public ApplicationService(IApplicationRepository applicationRepository)
+        {
+            _applicationRepository = applicationRepository;
+        }
+        
+        /// <inheritdoc/>
+        public async Task<(bool Bool, ServiceError ServiceError)> ValidateDataTypeForApp(string org, string appId, string dataType)
```
Comment (acn-sbuad) [inline src/Storage/Services/ApplicationService.cs:25]:
```suggestion
        public async Task<(bool IsValid, ServiceError ServiceError)> ValidateDataTypeForApp(string org, string appId, string dataType)
```

File: src/Storage/Services/DataService.cs [hunk 4]
```diff
@@ -45,5 +50,41 @@ public async Task StartFileScan(Instance instance, DataType dataType, DataElemen
                 await _fileScanQueueClient.EnqueueFileScan(serialisedRequest, ct);
             }
         }
+
+        /// <inheritdoc/>
+        public async Task<(string String, ServiceError ServiceError)> GenerateSha256Hash(string org, Guid instanceGuid, Guid dataElementId)
```
Comment (acn-sbuad) [inline src/Storage/Services/DataService.cs:55]:
```suggestion
        public async Task<(string BlobHash, ServiceError Error)> GenerateSha256Hash(string org, Guid instanceGuid, Guid dataElementId)
```

File: src/Storage/Services/DataService.cs [hunk 5]
```diff
@@ -45,5 +50,41 @@ public async Task StartFileScan(Instance instance, DataType dataType, DataElemen
...
+                return (null, new ServiceError(404, $"DataElement not found, dataElementId: {dataElementId}"));
+            }
+
+            Stream filestream = await _dataRepository.ReadDataFromStorage(org, dataElement.BlobStoragePath);
+            if (filestream == null || filestream.Length == 0)
+            {
+                return (null, new ServiceError(404, $"Failed reading file, dataElementId: {dataElementId}"));
+            }
+
+            return (CalculateSha256Hash(filestream), null);
```
Comment (acn-sbuad) [inline src/Storage/Services/DataService.cs:69]: are we sure errors cannot occur during the hashing? E.g. with empty files?

File: src/Storage/Services/IApplicationService.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    /// </summary>
+    public interface IApplicationService
+    {
+        /// <summary>
+        /// Upload file and save dataElement
+        /// </summary>
+        /// <param name="org">The application owner id.</param>
+        /// <param name="appId">The id of the application.</param>
+        /// <param name="dataType">The data type identifier for the data being uploaded.</param>
+         Task<(bool Bool, ServiceError ServiceError)> ValidateDataTypeForApp(string org, string appId, string dataType);
```
Comment (acn-sbuad) [inline src/Storage/Services/IApplicationService.cs:17]: Fix names of tuple properties

File: src/Storage/Services/DataService.cs (hunk 5 above)
Comment (tba76) [reply src/Storage/Services/DataService.cs:69]: Filestream is checked not to be null or empty before this point. Any other errors should be 500 from controller.

File: src/Storage/Configuration/PlatformSettings.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        /// <summary>
+        /// Gets or sets the url for the Authorization API endpoint.
+        /// </summary>
+        public string ApiAuthorizationEndpoint { get; set; }
+
+        /// <summary>
+        /// Gets or sets the url for the Profile API endpoint
+        /// </summary>
+        public string ApiProfileEndpoint { get; set; }
```
[the remark this answers is not in the document]
Comment (tba76) [reply src/Storage/Configuration/PlatformSettings.cs]: Removed PlatformSettings.

File: src/Storage/Services/IApplicationService.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Altinn.Platform.Storage.Services
+{
+    /// <summary>
+    /// This interface describes the required methods and features of a application service implementation.
+    /// </summary>
+    public interface IApplicationService
+    {
+        /// <summary>
+        /// Upload file and save dataElement
```
Comment (SandGrainOne) [inline src/Storage/Services/IApplicationService.cs:12]: - Wrong summary. 
- Maybe add a description for the return type?

File: src/Storage/Services/InstanceService.cs [hunk 9]
```diff
@@ -13,25 +16,91 @@ namespace Altinn.Platform.Storage.Services
...
         {
             _instanceRepository = instanceRepository;
+            _dataService = dataService;
+            _applicationService = applicationService;
             _instanceEventService = instanceEventService;
         }
 
         /// <inheritdoc/>
-        public async Task<SignDocument> CreateSignDocument(int instanceOwnerPartyId, Guid instanceGuid, SignRequest signRequest)
+        public async Task<(bool Created, ServiceError ServiceError)> CreateSignDocument(int instanceOwnerPartyId, Guid instanceGuid, SignRequest signRequest, int userId)
```
Comment (SandGrainOne) [inline src/Storage/Services/InstanceService.cs:35]: We could consider moving this to a "Sign service"

Comment (tba76) [reply src/Storage/Services/InstanceService.cs:35]: It's no problem to call it SignService but we use instanceRepository so it was clean to call it InstanceService. 

UserId is userId of the authenticated user. Users of the api can send in whatever they want in requestbody but like this we can log who did it.

File: src/Storage/Services/IApplicationService.cs (hunk 8 above)
Comment (tba76) [reply src/Storage/Services/IApplicationService.cs]: Done

File: src/Storage/Controllers/SignController.cs [hunk 10]
```diff
@@ -41,8 +40,20 @@ public SignController(IInstanceService instanceService)
...
+                {
+                    return Problem("The 'UserId' parameter must be defined for signee.", null, 400);
+                }
+
+                await _instanceService.CreateSignDocument(instanceOwnerPartyId, instanceGuid, signRequest, User.GetUserIdAsInt().Value);
+                return StatusCode(201, "SignDocument is created");
+            } 
+            catch (Exception e)
+            {
+                return Problem(e.Message, null, 500);
```
[the remark this answers is not in the document]
Comment (tba76) [reply src/Storage/Controllers/SignController.cs]: Removed try/catch

File: src/Storage/Altinn.Platform.Storage.csproj [hunk 11]
```diff
@@ -9,8 +9,12 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="Altinn.Common.AccessToken" Version="2.0.0" />
+    <PackageReference Include="Altinn.Common.AccessTokenClient" Version="1.1.3" />
+    <PackageReference Include="Altinn.Platform.Models" Version="1.2.0" />
     <PackageReference Include="Altinn.Common.AccessToken" Version="2.0.0" />
     <PackageReference Include="Altinn.Platform.Storage.Interface" Version="3.20.0" />
```
Comment (acn-sbuad) [inline src/Storage/Altinn.Platform.Storage.csproj:16]:
```suggestion
    <PackageReference Include="Altinn.Platform.Storage.Interface" Version="3.21.0" />
```

File: src/Storage/Controllers/SignController.cs [hunk 12]
```diff
@@ -41,7 +40,12 @@ public SignController(IInstanceService instanceService)
         [Produces("application/json")]
         public async Task<ActionResult> Sign([FromRoute] int instanceOwnerPartyId, [FromRoute] Guid instanceGuid, [FromBody] SignRequest signRequest)
         {
-            await _instanceService.CreateSignDocument(instanceOwnerPartyId, instanceGuid, signRequest);
+            if (string.IsNullOrEmpty(signRequest.Signee.UserId))
```
Comment (acn-sbuad) [inline src/Storage/Controllers/SignController.cs:43]:
```suggestion
            if (string.IsNullOrEmpty(signRequest.Signee?.UserId))
```

File: src/Storage/Controllers/SignController.cs [hunk 13]
```diff
@@ -41,7 +40,12 @@ public SignController(IInstanceService instanceService)
         [Produces("application/json")]
         public async Task<ActionResult> Sign([FromRoute] int instanceOwnerPartyId, [FromRoute] Guid instanceGuid, [FromBody] SignRequest signRequest)
         {
-            await _instanceService.CreateSignDocument(instanceOwnerPartyId, instanceGuid, signRequest);
+            if (string.IsNullOrEmpty(signRequest.Signee?.UserId))
```
Comment (acn-sbuad) [inline src/Storage/Controllers/SignController.cs:43]:
```suggestion
            if (string.IsNullOrEmpty(signRequest?.Signee?.UserId))
```

File: src/Storage/Models/ServiceError.cs [hunk 14]
```diff
@@ -0,0 +1,28 @@
+namespace Altinn.Platform.Storage.Models
+{
+    /// <summary>
+    /// A class representing a service error object used to transfere error information from service to controller.
```
Comment (acn-sbuad) [inline src/Storage/Models/ServiceError.cs:4]:
```suggestion
    /// A class representing a service error object used to transfer error information from service to controller.
```

File: src/Storage/appsettings.json [hunk 15]
```diff
@@ -29,6 +29,7 @@
   },
   "PlatformSettings": {
     "ApiAuthorizationEndpoint": "http://localhost:5050/authorization/api/v1/",
+    "ApiProfileEndpoint": "http://localhost:5101/profile/api/v1/",
```
Comment (acn-sbuad) [inline src/Storage/appsettings.json:32]: should be deleted

**Review by acn-sbuad — APPROVED** [review]
🎈

**Pull request merged**